swarm/src/stream_protocol.rs

```diff
@@ -0,0 +1,95 @@
+use either::Either;
+use std::fmt;
+use std::hash::{Hash, Hasher};
+use std::sync::Arc;
+
+/// Identifies a protocol for a stream.
+///
+/// libp2p nodes use stream protocols to negotiate what to do with a newly opened stream.
+/// Stream protocols are string-based and must start with a forward slash: `/`.
+#[derive(Debug, Clone, Eq)]
+pub struct StreamProtocol {
```

Comment:
To be nit-picky, one could as well use `StreamProtocol` to upgrade a connection, e.g. with a security upgrade like noise. That said, I don't have a better naming suggestion.

Another nit-pick, `StreamProtocolName` would be in my eyes more descriptive, though more verbose. I am fine either way.

Reply:
> To be nit-picky, one could as well use `StreamProtocol` to upgrade a connection, e.g. with a security upgrade like noise. That said, I don't have a better naming suggestion.

I did think about that, however I consider this to be the first step towards separating the use of multistream-select for transport upgrades from stream upgrades.

With increased usage of QUIC, transport upgrades will no longer be needed but stream upgrades will remain. Thus I figured it is good to create an abstraction for stream upgrades that is separate. Note that `StreamProtocol` lives in `libp2p-swarm` whereas all transport upgrades live in `libp2p-core`.

> Another nit-pick, `StreamProtocolName` would be in my eyes more descriptive, though more verbose. I am fine either way.

It is a bit too long for my taste. In discussions, we never talk about a protocol _name_, we talk about stream protocols (vs multiaddr protocols for example).

I think I'll leave it as is.